File: services/transcriber-rs/src/storage.rs
use anyhow::{Context, Result};
use s3::creds::Credentials;
use s3::{Bucket, Region};
use std::path::Path;
use tokio::fs::File;
use tokio::io::AsyncWriteExt;
use tracing::info;

use crate::config::Config;

pub struct Storage {
    bucket: Box<Bucket>,
}

impl Storage {
    pub fn new(config: &Config) -> Result<Self> {
        let region = Region::Custom {
            region: "us-east-1".to_string(),
            endpoint: format!(
                "{}://{}",
                if config.minio_use_ssl { "https" } else { "http" },
                config.minio_endpoint
            ),
        };

        let credentials = Credentials::new(
            Some(&config.minio_access_key),
            Some(&config.minio_secret_key),
            None,
            None,
            None,
        )?;

        let bucket = Bucket::new(&config.minio_bucket, region, credentials)?
            .with_path_style();

        Ok(Self { bucket })
    }

    pub async fn download_file(&self, object_key: &str, local_path: &Path) -> Result<()> {
        info!("Downloading {} to {:?}", object_key, local_path);

        // Get object from S3/MinIO
        let response = self
            .bucket
            .get_object(object_key)
            .await
            .context("Failed to get object from MinIO")?;

        // Write to local file
        let mut file = File::create(local_path)
            .await
            .context("Failed to create local file")?;

        file.write_all(response.bytes())
            .await
            .context("Failed to write file")?;

        file.flush().await?;

        info!("Downloaded {} bytes", response.bytes().len());
        Ok(())
    }

    pub fn normalize_object_key<'a>(&self, file_url: &'a str, bucket_name: &str) -> &'a str {
        // Handle various URL formats:
        // - "recordings/meeting-123/user-456.ogg"
        // - "meeting-123/user-456.ogg"
        // - Full URL with bucket prefix
        file_url
            .strip_prefix(&format!("{}/", bucket_name))
            .unwrap_or(file_url)
    }
}

File: services/transcriber-py/app/services/storage.py
import asyncio
import logging
from pathlib import Path

from minio import Minio

logger = logging.getLogger(__name__)


class StorageService:
    """MinIO S3 storage client."""

    def __init__(
        self,
        endpoint: str,
        access_key: str,
        secret_key: str,
        bucket: str,
        secure: bool = False,
    ):
        self.client = Minio(
            endpoint,
            access_key=access_key,
            secret_key=secret_key,
            secure=secure,
        )
        self.bucket = bucket

    def normalize_object_key(self, file_url: str) -> str:
        """Normalize file URL to object key.

        Handles various formats:
        - "recordings/meeting-123/user-456.ogg"
        - "meeting-123/user-456.ogg"
        """
        if file_url.startswith(f"{self.bucket}/"):
            return file_url[len(self.bucket) + 1 :]
        return file_url

    async def download_file(self, object_key: str, local_path: Path) -> None:
        """Download file from MinIO to local path."""
        object_key = self.normalize_object_key(object_key)
        logger.info(f"Downloading {object_key} to {local_path}")

        # minio-py is sync, run in executor
        loop = asyncio.get_event_loop()
        await loop.run_in_executor(
            None,
            lambda: self.client.fget_object(self.bucket, object_key, str(local_path)),
        )

        logger.info(f"Downloaded {object_key}")
